drs_downloader/manager.py

```diff
@@ -0,0 +1,383 @@
+        Args:
+            file: destination file
+            hash_method: instantiated hash_method
+        """
+        self._file = file
+        self._hash_method = hash_method
+
+    def read(self, size):
+        buffer = self._file.read(size)
+        self._hash_method.update(buffer)
+        return buffer
+
+    def __getattr__(self, attr):
+        return getattr(self._file, attr)
+
+
+class DrsAsyncManager(DrsManager):
+    """Manage DRSClient workload with asyncio threads, display progress.
+
+    """
+
+    def __init__(self, drs_client: DrsClient,
+                 show_progress: bool = True,
+                 part_size: int = DEFAULT_PART_SIZE,
+                 max_simultaneous_object_retrievers=DEFAULT_MAX_SIMULTANEOUS_OBJECT_RETRIEVERS,
+                 max_simultaneous_downloaders=DEFAULT_MAX_SIMULTANEOUS_DOWNLOADERS,
+                 max_simultaneous_part_handlers=DEFAULT_MAX_SIMULTANEOUS_PART_HANDLERS):
+        """
+
+        Args:
+            drs_client: the client that will interact with server
+            show_progress: show progress bars
+            part_size: tweak to optimize workload
+            max_simultaneous_object_retrievers: tweak to optimize workload
+            max_simultaneous_downloaders: tweak to optimize workload
+            max_simultaneous_part_handlers: tweak to optimize workload
+        """
+        # """Implements abstract constructor."""
+        super().__init__(drs_client=drs_client)
+        self.max_simultaneous_object_retrievers = max_simultaneous_object_retrievers
+        self.max_simultaneous_downloaders = max_simultaneous_downloaders
+        self.max_simultaneous_part_handlers = max_simultaneous_part_handlers
+        self.disable = not show_progress
+        self.part_size = part_size
+
+    @staticmethod
+    def _parts_generator(size: int, start: int = 0, part_size: int = None) -> Iterator[Tuple[int, int]]:
+        """Determine the start,size for each part
+
+        Args:
+            size: size of file
+            start: offset into file 0 based
+            part_size: desired part size
+
+        Returns:
+            iterator of tuple start, size
+        """
+        while size - start > part_size:
+            yield start, start + part_size
+            start += part_size + 1
+            # start += part_size
+        yield start, size
+
+    async def _run_download_parts(self, drs_object: DrsObject, destination_path: Path) -> DrsObject:
+        """Determine number of parts for signed url and create tasks for each part, run them in batches.
+
+        Args:
+            drs_object: Information about a bucket object
+
+        Returns:
+            list of paths to files for each part, in order.
+        """
+        # create a list of parts
+        parts = []
+        for start, size in self._parts_generator(size=drs_object.size, part_size=self.part_size):
+            parts.append((start, size, ))
+
+        if len(parts) > 1000:
```

Comment:
This is not the number of open files at any given moment, this is just measuring the number of open
files total that will be opened for each file's download.